utils/utils.go

```diff
@@ -111,13 +112,31 @@ func CleanUpNamespace(args *skel.CmdArgs, logger *logrus.Entry) error {
 
 		if devErr == nil {
 			fmt.Fprintf(os.Stderr, "Calico CNI deleting device in netns %s\n", args.Netns)
-			err := ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
-				_, err := ip.DelLinkByNameAddr(args.IfName, netlink.FAMILY_V4)
-				return err
-			})
+			// Deleting the veth has been seen to hang on some kernel version. Timeout the command if it takes too long.
+			ch := make(chan error, 1)
 
-			if err != nil {
-				return err
+			go func() {
+				err := ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+					_, err := ip.DelLinkByNameAddr(args.IfName, netlink.FAMILY_V4)
+					return err
+				})
+
+				if err != nil {
+					ch <- err
+				} else {
+					ch <- nil
+				}
+			}()
+
+			select {
+			case err := <-ch:
+				if err != nil {
+					return err
+				} else {
+					fmt.Fprintf(os.Stderr, "Calico CNI deleted device in netns %s\n", args.Netns)
+				}
+			case <-time.After(5 * time.Second):
+				fmt.Fprintf(os.Stderr, "Calico CNI timed out deleting device in netns %s\n", args.Netns)
```

Comment:
Did you consider returning an error in this scenario?

I suppose it's probably OK, but might we end up leaking devices this way?